include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,371 @@
+/* boost random/nierderreiter_base2.hpp header file
+ *
+ * Copyright Justinas Vygintas Daugmaudis 2010-2018
+ * Distributed under the Boost Software License, Version 1.0. (See
+ * accompanying file LICENSE_1_0.txt or copy at
+ * http://www.boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_RANDOM_NIEDERREITER_BASE2_HPP
+#define BOOST_RANDOM_NIEDERREITER_BASE2_HPP
+
+#include <boost/random/detail/niederreiter_base2_data.hpp>
+#include <boost/random/detail/gray_coded_qrng_base.hpp>
+#include <boost/random/detail/config.hpp>
+#include <boost/random/detail/operators.hpp>
+
+#include <limits>
+#include <boost/cstdint.hpp>
+
+#include <boost/dynamic_bitset.hpp>
+#include <boost/multi_array.hpp>
+
+//!\file
+//!Describes the quasi-random number generator class template niederreiter_base2_engine.
+//!
+//!\b Note: it is especially useful in conjunction with class template uniform_real.
+
+namespace boost {
+namespace random {
+
+/** @cond */
+namespace detail {
+namespace nb2 {
+
+// Return the base 2 logarithm for a given bitset v
+template <typename Block, typename Allocator>
+inline typename boost::dynamic_bitset<Block, Allocator>::size_type
+bitset_log2(const boost::dynamic_bitset<Block, Allocator>& v)
+{
+  typedef boost::dynamic_bitset<Block, Allocator> bitset_t;
+  typedef typename bitset_t::size_type size_type;
+
+  if (v.none())
+    throw std::invalid_argument("bitset_log2");
+
+  size_type up  = v.size() - 1;
+  size_type low = v.find_next(0);
+
+  // Binary lookup for the most significant set bit
```

Comment:
Alas, this function is necessary, since pb_degree does not increase monotonically. It is quite possible for kj (the previous pb_degree value) to be bigger than the next computed pb_degree. Try and printout the cases, when kj > pb_degree.

As for the reverse scan, you are right. I somehow didn't think of it. Done.

Nevertheless, I will note that this binary search technique is not totally without merit. It effectively cuts off all the lower bits if any upper bits are already set.

Reply:
I have tested the performance of the revscan approach. I have to say that your reverse scan idea works slightly faster than bin search. Both revscan and bin search beat silly 

```C++
  size_type low = 0;
  size_type p = 0;

  // Linear lookup for the most significant set bit
  while (p != DynamicBitset::npos)
  {
    low = p;
    p = v.find_next(low);
  }

  return low;
```

hands down.

Timings on my machine for 10000000 random bitsets are like this:
bin_search: 2.35857s
rev_scan: 2.33807s
lin_search: 4.01772s